erts/emulator/beam/erl_init.c

```diff
@@ -762,6 +806,110 @@ static void ethr_ll_free(void *ptr)
     erts_free(ERTS_ALC_T_ETHR_LL, ptr);
 }
 
+#if defined(__linux__) && defined(__x86_64__) && defined(__LP64__)
+static int
+is_transparent_hugepage_madvise_enabled(void)
+{
+    char value[10];
+    const char *path;
+    FILE *fp;
+    int n;
+
+    path = "/sys/kernel/mm/transparent_hugepage/enabled";
+    if ((fp = fopen(path, "r")) == NULL) {
+        erts_fprintf(stderr, "cannot open '%s', is sysfs mounted?\n", path);
+        return 0;
+    }
+
+    n = fscanf(fp, "%*s %9s %*s", value); /* strlen("[madvise]") is 9 */
+    fclose(fp);
+    if (n != 1) {
+        erts_fprintf(stderr, "could not parse '%s'\n", path);
+        return 0;
+    }
+
+    return (sys_strcmp(value, "[madvise]") == 0);
+}
+
+static int
+maybe_madvise_hugepage(char *from, char *to)
+{
+    extern char etext; /* &etext is the first address above the .text segment */
+    Uint hugepage_size;
+
+    /* Establish the invariant from <= etext <= to */
+
+    if (to < &etext) {
+        /* Not there yet, continue. */
+        return 0;
+    }
+
+    if (from > &etext) {
+        /* Gone too far, stop. */
+        return 1;
+    }
+
+    /*
+     * According to the Linux kernel documentation, in order to
+     * allocate 2MiB pages immediately, memory regions must be 2MiB
+     * aligned.
+     *
+     * The build should have instructed the linker to 2MiB align
+     * segments, if they are not, something has gone wrong.
```

Comment:
The `etext` symbol is provided by the linker but you are right about there being no guarantee of 2MiB section alignment.  That is my oversight, I pass a flag to the linker to increase the alignment granularity but I did not include that in this change.

Here is what things probably look like by default, using a simple "hello, world" program on a vanilla Linux distribution

```
fedora {25} cc -o hello-4kib hello.c
fedora {26} readelf -l hello-4kib 

Elf file type is EXEC (Executable file)
Entry point 0x401040
There are 13 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
                 0x00000000000002d8 0x00000000000002d8  R      0x8
  INTERP         0x0000000000000318 0x0000000000400318 0x0000000000400318
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000000000508 0x0000000000000508  R      0x1000
  LOAD           0x0000000000001000 0x0000000000401000 0x0000000000401000
                 0x0000000000000161 0x0000000000000161  R E    0x1000
...
```

And here is what they look like when we override the page size

```
fedora {27} cc -o hello-2mib -Wl,-zcommon-page-size=2097152 hello.c
fedora {28} readelf -l hello-2mib 

Elf file type is EXEC (Executable file)
Entry point 0x600040
There are 13 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
                 0x00000000000002d8 0x00000000000002d8  R      0x8
  INTERP         0x0000000000000318 0x0000000000400318 0x0000000000400318
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000000000508 0x0000000000000508  R      0x1000
  LOAD           0x0000000000200000 0x0000000000600000 0x0000000000600000
                 0x0000000000000161 0x0000000000000161  R E    0x1000
...
```

You can see that the text segment starts at 0x60000 instead of 0x401000, which is what we need.  This flag is something that can probably be added by the `configure` script contingent on all of the other THP stuff being present.